### PR TITLE
Adapt  NPU fusion operators to Qwen2.5-VL model

### PR DESCRIPTION
**Pull request by frozenleaves**
This PR is aimed at adapting the Ascend NPU fusion operators for the `Qwen2.5-VL` series models to enhance their performance on the NPU. Currently, this PR only makes **non-compatible modifications for the NPU** and does not meet the conditions for merging into the mainline. **It is only applicable to the joint innovation version for restricted customers**. If any bugs are encountered, communication and updates can be made at any time.

The modifications in this PR include enabling support for three NPU fusion operators: NPU Flash Attention, NPU RMSNorm, and NPU SwiGLU. Currently, it **only supports the Qwen2.5-VL series models**, and large-scale verification has only been conducted on the Qwen2.5-VL-7B model. Other series of the Qwen2.5-VL models (such as 32B) have not been verified, and accuracy and performance are not guaranteed.

### CODE REVIEW

File: src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py [hunk 1]
```diff
@@ -69,8 +70,10 @@ def __init__(self, config, bias: bool = False):
         self.act_fn = ACT2FN[config.hidden_act]
 
     def forward(self, hidden_state):
-        return self.down_proj(self.act_fn(self.gate_proj(hidden_state)) * self.up_proj(hidden_state))
-
+        # return self.down_proj(self.act_fn(self.gate_proj(hidden_state)) * self.up_proj(hidden_state))
+        return self.down_proj(
+            torch_npu.npu_swiglu(torch.cat((self.gate_proj(hidden_state), self.up_proj(hidden_state)), dim=-1), dim=-1)
+        )
```
Comment (vasqu) [inline src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py:76]: Does `self.down_proj(self.act_fn(self.gate_proj(hidden_state)) * self.up_proj(hidden_state)` not work or is it that `torch_npu.npu_swiglu` is faster?

File: src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py [hunk 2]
```diff
@@ -120,11 +123,12 @@ def __init__(self, hidden_size, eps=1e-6):
         self.variance_epsilon = eps
 
     def forward(self, hidden_states):
-        input_dtype = hidden_states.dtype
-        hidden_states = hidden_states.to(torch.float32)
-        variance = hidden_states.pow(2).mean(-1, keepdim=True)
-        hidden_states = hidden_states * torch.rsqrt(variance + self.variance_epsilon)
-        return self.weight * hidden_states.to(input_dtype)
+        return torch_npu.npu_rms_norm(hidden_states, self.weight, epsilon=self.variance_epsilon)[0]
```
Comment (vasqu) [inline src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py:126]: Same here

File: src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py [hunk 3]
```diff
@@ -156,6 +160,110 @@ def apply_rotary_pos_emb_flashatt(
...
+        q, k = apply_rotary_pos_emb_custom(q, k, cos, sin)
+
+        head_num = q.shape[1]
+        scale_value = 1.0 / math.sqrt(q.size(-1))
+
+        if not hasattr(self, 'cached_seq_len') or self.cached_seq_len != tuple(cu_seqlens[1:].tolist()):
+            self.cached_seq_len = tuple(cu_seqlens[1:].tolist())
+            self.actual_seq_len = tuple(cu_seqlens[1:].cpu().numpy().tolist())
+
+        attn_output = torch_npu.npu_fusion_attention(
```
Comment (vasqu) [inline src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py:246]: We support npu flash attention already, see https://github.com/huggingface/transformers/blob/main/src/transformers/integrations/npu_flash_attention.py

You can access it by loading with `attn_implementation="flash_attention_2"` (it will automatically look for the proper flash attn installation, including npu).

**Review by vasqu — COMMENTED** [review]
Did you encounter any issues when using NPU? Imo, everything should already work and some NPU-specific optimizations (flash attention) are already available. (And the swiglu, rmsnorm don't seem worth it as we rather have an explicit version here atm)

It also seems that you are on an older version as the attention implementation is outdated - we now use an attn interface wrapper which handles the switch between using your requested attn flavor.

**Comment by vasqu** [comment]
Sorry, just read the `and does not meet the conditions for merging into the mainline` part :sweat: I'm confused why this PR was opened then? Maybe put it as draft for now?

My comments for flash attn on npu stand tho - it should work on main already!
